src/rgw/driver/sfs/multipart.cc

```diff
@@ -13,301 +13,616 @@
-
-  if (mp->state == sfs::MultipartUpload::State::ABORTED) {
-    lsfs_dout(dpp, 10) << "multipart with upload_id " << mp->upload_id
-                       << " has been aborted." << dendl;
-    return -ERR_NO_SUCH_UPLOAD;
-  } else if (mp->state == sfs::MultipartUpload::State::DONE) {
-    lsfs_dout(dpp, 10) << "multipart with upload_id " << mp->upload_id
-                       << " has been completed." << dendl;
+  lsfs_dout(dpp, 10) << fmt::format(
+                            "upload_id: {}, accounted_size: {}, tag: {}, "
+                            "owner: {}, olh_epoch: {}"
+                            ", target_obj: {}",
+                            upload_id, accounted_size, tag,
+                            owner.get_display_name(), olh_epoch,
+                            target_obj->get_key()
+                        )
+                     << dendl;
+  lsfs_dout(dpp, 10) << "part_etags: " << part_etags << dendl;
+
+  sfs::sqlite::SQLiteMultipart mpdb(store->db_conn);
+  bool duplicate = false;
+  auto res = mpdb.mark_complete(upload_id, &duplicate);
+  if (!res) {
+    lsfs_dout(dpp, 10) << fmt::format(
+                              "unable to find on-going multipart upload id {}",
+                              upload_id
+                          )
+                       << dendl;
     return -ERR_NO_SUCH_UPLOAD;
+  } else if (duplicate) {
+    lsfs_dout(dpp, 10)
+        << fmt::format(
+               "multipart id {} already completed, returning success!",
+               upload_id
+           )
+        << dendl;
+    return 0;
   }
 
-  auto parts = mp->get_parts();
+  auto parts = mpdb.get_parts(upload_id);
   if (parts.size() != part_etags.size()) {
     return -ERR_INVALID_PART;
   }
 
-  mp->aggregate();
+  auto mp = mpdb.get_multipart(upload_id);
+  ceph_assert(mp.has_value());
+  ceph_assert(mp->upload_id == upload_id);
+  ceph_assert(mp->state == sfs::MultipartState::COMPLETE);
+
+  // validate parts & build final etag
 
-  MD5 hash;
+  // we can only have at most 10k parts
+  if (part_etags.size() > 10000) {
+    return -ERR_INVALID_PART;
+  }
 
-  ceph_assert(target_obj);
-  ceph_assert(target_obj->get_name() == mp->objref->name);
-  sfs::ObjectRef outobj = bucketref->get_or_create(target_obj->get_key());
-  std::filesystem::path outpath =
-      store->get_data_path() / outobj->get_storage_path();
-  // ensure directory structure exists
-  std::filesystem::create_directories(outpath.parent_path());
+  std::map<int, sqlite::DBMultipartPart> to_complete;
+  std::map<int, sqlite::DBMultipartPart> parts_map;
+  for (const auto& p : parts) {
+    parts_map[p.part_num] = p;
+  }
 
-  ofstream out{outpath, std::ios::binary | std::ios::app};
+  ETagBuilder hash;
+  uint64_t expected_size = 0;
+  // by nature, the `part_etags` map is an ordered container; all parts are
+  // already provided sorted.
+  for (auto it = part_etags.cbegin(); it != part_etags.cend(); ++it) {
```

Comment:
I don't know of a good way to check whether the element being handled is the last element, if using the construct you are suggesting. As least, with the current `it` construct, we can check whether the iterator is at a given distance from `part_etags.cend()`.

Let me know if you know of a good alternative. :)

Reply:
Makes sense. Suggest adding a const though if it doesn't need to be modified. `const auto& it` perhaps.

Reply:
Out of curiosity, wouldn't that be the end result of the `auto` as is, given we're asking for a `const` iterator anyway (i.e., `cbegin()` vs `begin()`)?

Reply:
It wound not actually work, because `it` must be modifiable by the loop :). 
Was a bit curious myself and asked godbolt how various loop constructs and auto do types.

<details>
<summary>compiler explorer c++ insights</summary>

```
#include <iostream>
#include <vector>

int main(int argc, char** argv) {
    std::vector foo = {"a", "b", "c"};

    for (auto it = foo.cbegin(); it != foo.cend(); ++it) {
        std::cout << *it;
    }

    for (const auto& it : foo) {
        std::cout << it;
    }
  
    for (auto it : foo) {
        std::cout << it;
    }
  
    for (auto& it : foo) {
        std::cout << it;
    }
}
```


```
// Type your code here, or load an example.
#include <iostream>
#include <vector>

int main(int argc, char ** argv)
{
  std::vector<const char *, std::allocator<const char *> > foo = std::vector<const char *, std::allocator<const char *> >{std::initializer_list<const char *>{"a", "b", "c"}, std::allocator<const char *>()};
  for(__gnu_cxx::__normal_iterator<const char *const *, std::vector<const char *, std::allocator<const char *> > > it = foo.cbegin(); !__gnu_cxx::operator==(it, foo.cend()); it.operator++()) {
    std::operator<<(std::cout, it.operator*());
  }
  
  {
    std::vector<const char *, std::allocator<const char *> > & __range1 = foo;
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __begin1 = __range1.begin();
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __end1 = __range1.end();
    for(; !__gnu_cxx::operator==(__begin1, __end1); __begin1.operator++()) {
      const char *const & it = __begin1.operator*();
      std::operator<<(std::cout, it);
    }
    
  }
  {
    std::vector<const char *, std::allocator<const char *> > & __range1 = foo;
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __begin1 = __range1.begin();
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __end1 = __range1.end();
    for(; !__gnu_cxx::operator==(__begin1, __end1); __begin1.operator++()) {
      const char * it = __begin1.operator*();
      std::operator<<(std::cout, it);
    }
    
  }
  {
    std::vector<const char *, std::allocator<const char *> > & __range1 = foo;
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __begin1 = __range1.begin();
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __end1 = __range1.end();
    for(; !__gnu_cxx::operator==(__begin1, __end1); __begin1.operator++()) {
      const char *& it = __begin1.operator*();
      std::operator<<(std::cout, it);
    }
    
  }
  return 0;
}
```

</details>